### PR TITLE
refactor: Credentials Validation

### PR DESCRIPTION
**Pull request by nkomonen-amazon**
This refactors the credentials validation code that previously had multiple validation functions for each credentials key.

This refactor changes this to use a switch case instead so that all validations of a certain type are contained within
the same function. This improves readability and reduces noise.

### Additional

- Name changes

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/credentials/sharedCredentialsValidation.ts [hunk 1]
```diff
@@ -6,89 +6,127 @@
...
- * at once.
+ * Validates all credentials values from the given input
+ *
+ * @returns Returns the same shape as the input, but the value of each
+ * key is an error message if the original value is invalid,
+ * otherwise undefined
  */
-export type StaticCredentialsErrorResult = {
-    [k in keyof StaticCredentialsProfileData]: string | undefined
+export function getCredentialsErrors(
```
Comment (nkomonen-amazon) [inline src/credentials/sharedCredentialsValidation.ts:34]: This is the head of the change, it validates an object with credentials data. Internally, this calls individual validation functions.

File: src/credentials/sharedCredentialsValidation.ts [hunk 2]
```diff
@@ -6,89 +6,123 @@
...
+ * otherwise undefined.
+ *
+ * If there are no errors at all, undefined is returned
  */
-export type StaticCredentialsErrorResult = {
-    [k in keyof StaticCredentialsProfileData]: string | undefined
+export function getCredentialsErrors(
+    data: CredentialsData,
+    validateFunc: GetCredentialError = getCredentialError
+): CredentialsData | undefined {
```
[the remark this answers is not in the document]
Comment (nkomonen-amazon) [reply src/credentials/sharedCredentialsValidation.ts:33]: I created a new type `CredentialsErrors` which improves readability but maintains the same shape

File: src/credentials/sharedCredentialsValidation.ts [hunk 3]
```diff
@@ -6,89 +6,123 @@
...
+ *
+ * This function assumes there is a value to evaluate, if not
+ * it returns no error.
+ */
+export const getCredentialFormatError: GetCredentialError = (key: CredentialsKey, value: string | undefined) => {
+    if (!value) {
+        /** Empty values should be validated in {@link getCredentialEmptyError} */
+        getLogger().debug('getCredentialFormatError() called with empty value for key "%s"', key)
+        return
     }
```
[the remark this answers is not in the document]
Comment (nkomonen-amazon) [reply src/credentials/sharedCredentialsValidation.ts:71]: I'm not sure what you mean. Add the logic for handling an empty `value` to `getAccessKeyIdFormatError()`?

File: src/credentials/types.ts [hunk 4]
```diff
@@ -24,15 +24,18 @@ export const SharedCredentialsKeys = {
...
+
+export type CredentialsKey = (typeof SharedCredentialsKeys)[keyof typeof SharedCredentialsKeys]
+
+// /**
+//  * The required keys for a static credentials profile
+//  *
+//  * https://docs.aws.amazon.com/sdkref/latest/guide/feature-static-credentials.html
+//  */
+export type StaticCredentialsProfileKeysOptional = Pick<CredentialsData, 'aws_access_key_id' | 'aws_secret_access_key'>
+export type StaticCredentialsProfileKeys = Required<StaticCredentialsProfileKeysOptional>
```
[the remark this answers is not in the document]
Comment (nkomonen-amazon) [reply src/credentials/types.ts:38]: I made related fixes to this in my current work. I'll come back to this in the next PR

File: src/credentials/types.ts [hunk 5]
```diff
@@ -24,15 +24,18 @@ export const SharedCredentialsKeys = {
...
+/** An object that has only credentials data */
+export type CredentialsData = Partial<Record<CredentialsKey, string>>
+
+export type CredentialsKey = (typeof SharedCredentialsKeys)[keyof typeof SharedCredentialsKeys]
+
+// /**
+//  * The required keys for a static credentials profile
+//  *
+//  * https://docs.aws.amazon.com/sdkref/latest/guide/feature-static-credentials.html
+//  */
```
Comment (justinmk3) [inline src/credentials/types.ts:36]: double-comment

**Review by justinmk3 — APPROVED** [review]

**Pull request merged**